Fix stock status saving for variable products on bulk and quick edit

Konamiman opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes related to stock status for variable products on bulk and quick edit.

* **For bulk edit:** Even if stock status was left as "No change", the status of all variations was being changed to whatever the status of the product was before it was converted to variable. Now no change is performed when "No change" is selected, and all variations change to whatever is selected otherwise.

* **For quick edit:** A new "No change" option is added to the "Stock status" selector that will be preselected when the product is variable. Previously, whatever status the product had before being converted to variable was being shown, and that's the status that would be set when saving. Also, a "This will change the stock status of all variations" message is displayed before the selector.

![image](https://user-images.githubusercontent.com/937723/79217826-c58ba180-7e4f-11ea-8196-b7e6b3ec1a87.png)

#### Other improvements introduced:

* `WC_Unit_Test_Case` class gets two new utility methods: `login_as_role` and `login_as_administrator`.

* `WC_Helper_Product::create_simple_product` gets a new optional `$props` argument.

Closes #25843.

### How to test the changes in this Pull Request:

1. Create a variable product with two variations. Set stock status to "Out of stock" for one of them and to "On backorder" for the other.
2. Bulk edit the product, don't change anything and save. Verify that the stock status of the variations hasn't changed (without the fix, both would have changed to "In stock").
3. Bulk edit the product again, se the stock status selector to "In stock" and save. Verify that this time the status of both variations has changed.
4. Change the stock status of both variations to anything but "In stock" again.
5. Quick edit the product. Verify that the preselected value for the stock status selector is now "No change". Leave it that way and save. Verify that the stock status of the variations hasn't changed.
6. Quick edit the product again. Select "In stock" in the stock status selector. Verify that this time the status of both variations has changed.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - don' change the stock status of variations when bulk editing a variable product and leaving the "Stock status" selector as "No change".
> Enhancement - Add a "No change" option to the "Stock status" selector in quick edit, preselect it when the product being edited is a variable product.

Konamiman committed:
Fix code sniffer errors in WC_Admin_Post_Types.

Konamiman committed:
Refactor duplicate code & add rounding to price calculations in WC_Admin_Post_Types.

Two methods have been created:

- update_stock_status, replaces code that was duplicated in the
  quick_edit_save and bulk_edit_save methods.

- set_new_price, replaces code that was duplicated-ish in the
  bulk_edit_save for setting the new regular and sales prices
  (code was not identical but very similar).

Also, `round` is now used on sale price calculations that involve
multiplying by a percent, the same was as it was done already
to calculate the regular price.

Konamiman committed:
Fix code sniffer errors on quick-edit.js

Konamiman committed:
Fix stock status setting for variable products on bulk & quick edits.

For bulk edit: even if stock status was left as "No change", the
status of all variations was being changed to whatever the status
of the product was before it was converted to variable. Now
no change is performed when "No change" is selected, and all
variations change to whatever is selected otherwise.

For quick edit: a new "No change" option is added that will be
preselected when the product is variable. Previously, whatever
status the product had before being converted to variable was being
shown, and that's the status that would be set when saving.
Also, a "This will change the stock status of all variations"
message is displayed before the selector.

vedanshujain commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -986,6 +868,103 @@ public function add_display_post_states( $post_states, $post ) {
+	private function set_new_price( $product, $price_type ) {
+		// phpcs:disable WordPress.Security.NonceVerification.Recommended
+
+		if ( empty( $_REQUEST[ "change_{$price_type}_price" ] ) || ! isset( $_REQUEST[ "_{$price_type}_price" ] ) ) {
```
Just came across this commit, not a review and you probably already plan to do this, but since you are refactoring, it's probably better to shift `_REQUEST` calls in parent methods and pass the values in variables. This way method can be used outside of a `_REQUEST` context and unit testing will be slightly easier :).

Konamiman replied:
Hummm... yes and no. 🙂

I agree on extracting/wrapping `_REQUEST` somehow instead of using it directly, to ease unit testing. I also agree that methods should generally receive only the data they need. However in this particular case I prefer to access the request data directly instead of having it passed as parameters, otherwise I would have to extract the initial `if ( empty( ...` line out and that would add some duplication.

If at some point we really need to use this method outside of the context of a request, then sure, we can give it another shot and think how to make it completely request-free.

That said, I have created a new `request_data` method that just returns `_$REQUEST`, and changed the entire class to use it instead of the superglobal. Eventually we should find a way to do it consistently throughout the entire codebase, e.g. making it a method in an abstract base class.

Konamiman committed:
Extract $_REQUEST to a method in WC_Admin_Post_Types.

Create a new `request_data` method in WC_Admin_Post_Types that
just returns $_REQUEST. This is intended to ease unit testing,
as this method can be easily mocked to return test data.

Konamiman committed:
Fix code sniffer errors in a few unit test files.

Konamiman committed:
Add login_as_role and login_as_administrator to WC_Unit_Test_Case.

Those methods are a convenient replacement for
"this->factory->user->create". Tests that were using that to
simulate user login have been modified to use the new methods.

Konamiman committed:
Add unit test for WC_Admin_Post_Types::bulk_and_quick_edit_save_post.

The test added checks that stock status of variations when saving
a variable product is changed or not appropriately depending on
the request data supplied.

Konamiman committed:
Small fix in WC_Tests_Admin_Post_Types::bulk_and_quick_edit_stock_status_for_variable_product.

At some point the 'change_stock' key is assumed to be present
in the request data, but it might not. Fixed to test for existence
before using the value.

Konamiman committed:
Improve WC_Helper_Product::create_simple_product.

Add a new optional $props parameter containing the properties
to be set in the newly created product.

Konamiman committed:
Add unit test for WC_Admin_Post_Types::bulk_and_quick_edit_save_post.

The test added checks that the new regular or sale price is
appropriately set when bulk saving.

Konamiman commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -986,6 +868,107 @@ public function add_display_post_states( $post_states, $post ) {
+
+		if ( isset( $new_price ) && $new_price !== $old_price ) {
+			$price_changed = true;
+			if ( 'sale' === $price_type && empty( $new_price ) && '0' !== $new_price ) {
+				$new_price = '';
```
I've kept this logic from the pre-refactored code, but I'm not sure if it's really necessary; I can't think of any case where the price would end up being `''` (the previous `isset` check makes the `empty` check useless, right?)

vedanshujain replied:
I agree that it does not look necessary. maybe remove it?

vedanshujain commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -407,10 +422,17 @@ private function quick_edit_save( $post_id, $product ) {
+		} else {
+			$stock_status = $product->is_type( 'variable' ) ? null : 'instock';
+		}
+		$stock_amount = 'yes' === $manage_stock && isset( $request_data['_stock'] ) && is_numeric( wp_unslash( $request_data['_stock'] ) ) ? wc_stock_amount( wp_unslash( $request_data['_stock'] ) ) : '';
```
since you are already editing it, maybe also move this to before line 441?

vedanshujain commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -518,93 +529,10 @@ public function bulk_edit_save( $post_id, $product ) {
 
 		if ( $can_product_type_change_price ) {
-			$price_changed = false;
+			$regular_price_changed = $this->set_new_price( $product, 'regular' );
```
nice work refactoring this and making it testable ❤️

vedanshujain commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -616,24 +544,22 @@ public function bulk_edit_save( $post_id, $product ) {
 
 		if ( 'yes' === get_option( 'woocommerce_manage_stock' ) ) {
-			$change_stock = absint( $_REQUEST['change_stock'] );
+			$change_stock = empty( $request_data['change_stock'] ) ? 0 : absint( $request_data['change_stock'] );
```
`absint` will convert to 0 itself, no need for `empty` subclause i think

vedanshujain reviewed (changes requested):
Great work! Overall its looking pretty good and nice job in refactoring some messy logic! I have left some comments, but mostly they nitpick and code is testing well.

Konamiman commented on assets/js/admin/quick-edit.js:
```diff
@@ -1,135 +1,167 @@
+//phpcs:ignore
```
This is here just to ignore warnings on the next line, `/*global inlineEditPost, woocommerce_admin, woocommerce_quick_edit */`, which is not a regular comment but a directive; it doesn't affect the rest of the file

claudiosanches replied:
This is a `JS` file, so no PHPCS should inspect this file, so this sounds like some error on your code editor settings.

Konamiman commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -330,6 +337,10 @@ public function bulk_and_quick_edit_save_post( $post_id, $post ) {
 	 * @param WC_Product $product Product object.
 	 */
 	private function quick_edit_save( $post_id, $product ) {
+		// phpcs:disable WordPress.Security.NonceVerification.Recommended
```
This is a private method that is only used by `bulk_and_quick_edit_save_post`, which is doing a nonce verification before invoking it, so there's no need to verify nonces here.

Anyway, regarding the `disable`, now I see that I can just remove it and there are nor phpcs warnings 🤷🏻‍♂️

Konamiman commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -407,10 +422,17 @@ private function quick_edit_save( $post_id, $product ) {
+		$manage_stock            = ! empty( $request_data['_manage_stock'] ) && 'grouped' !== $product->get_type() ? 'yes' : 'no';
+		$backorders              = ! empty( $request_data['_backorders'] ) ? wc_clean( $request_data['_backorders'] ) : 'no';
+		$stock_status_in_request = ! empty( $request_data['_stock_status'] );
+		if ( $stock_status_in_request ) {
```
Really? I think the opposite, using that intermediate variable we are documenting what exactly are we looking for. But on the other hand it's true that doing a direct check would be more consistent with the rest of the code.

Konamiman commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -961,6 +880,122 @@ public function add_display_post_states( $post_states, $post ) {
+	 * @return array The $_REQUEST superglobal.
+	 */
+	protected function request_data() {
+		return $_REQUEST;
```
Not sure. The thing is that nonce verification is required only in the public method (where it's in place already), the rest of usages are inside private methods where there's no need to repeat the verification. I don't like the idea of adding additional logic just to make the code beautifier happy. Also `request_data` is intended to be a very tiny method which exists only to ease unit testing (now I think that perhaps `get_request_data` would be a better name, as in "this is a only a data getter").

claudiosanches replied:
We don't verify nonces on this class because WordPress already introduces some nonces and it get checked early.

Konamiman commented on includes/admin/views/html-quick-edit-product.php:
```diff
@@ -174,11 +175,15 @@
 					}
 					?>
 				</select>
+				<div class="quick-edit-warning" style="display:none">
```
Yes to both, that's why this is in place: https://github.com/woocommerce/woocommerce/pull/26174/files#diff-c3dfd13ac5cb3cdf7741fbad7cc433e6R62-R63

Konamiman commented on includes/admin/views/html-quick-edit-product.php:
```diff
@@ -174,11 +175,15 @@
 			<span class="input-text-wrap">
 				<select class="stock_status" name="_stock_status">
 					<?php
+					echo '<option value="" id="stock_status_no_change">' . esc_html__( '— No Change —', 'woocommerce' ) . '</option>';
```
This is done via JS (so that for variable products this one is selected, and for other products the old logic of selecting the current status is kept): https://github.com/woocommerce/woocommerce/pull/26174/files#diff-c3dfd13ac5cb3cdf7741fbad7cc433e6R64-R65 (`assets/js/admin/quick-edit.js` 64-65)

Konamiman committed:
Small adjustments in WC_Admin_Post_Types

- Remove no longer needed phpcs disablers.
- Rename method.
- Remove no longer needed logic.
- Simplify some logic.

Konamiman committed:
Remove unnecessary phpcs disabler in quick-edit.js

rrennick commented:
@vedanshujain Could you go through the comments above and clear the resolved items?

vedanshujain commented on includes/admin/class-wc-admin-post-types.php:
```diff
@@ -961,6 +874,118 @@ public function add_display_post_states( $post_states, $post ) {
+			case 2:
+				if ( $is_percentage ) {
+					$percent   = $price / 100;
+					$new_price = $old_price + ( round( $old_price * $percent, wc_get_price_decimals() ) );
```
[not a blocker] `round` is redundant here and on line 947 as we are doing it again on line 971 anyway.

vedanshujain reviewed (commented):
Changes look great and are testing well!! I have left a few comments, most of them are not blockers but one needs to be addressed before we can merge.

Konamiman commented on includes/admin/views/html-quick-edit-product.php:
```diff
@@ -174,11 +175,15 @@
 					}
 					?>
 				</select>
+				<div class="quick-edit-warning" style="display:none">
```
This is how it looks with `notice notice-warning`:

![image](https://user-images.githubusercontent.com/937723/87409729-a7a41900-c5c5-11ea-95de-87527cc6d69d.png)

I think it looks rather out of place. Maybe we could just bold and/or color the text?

Konamiman replied:
`bold` and `darkred` for comparison:

![image](https://user-images.githubusercontent.com/937723/87410046-1aad8f80-c5c6-11ea-8b9e-67596fac7752.png)

vedanshujain replied:
bold and dark red looks alright to me :+1:

Konamiman committed:
Fix error when `screen` not set in request, remove redundand `round`s.

Konamiman committed:
Rename `quick-edit-warning` style with `wc-` prefix, and add styling.

vedanshujain reviewed (approved):
LGTM!

claudiosanches committed:
Fixed docblocks

claudiosanches reviewed (approved):
Works great!

claudiosanches committed:
Merge branch 'master' into fix/25843
